### PR TITLE
Windows protocol handler

### PR DESCRIPTION
**Pull request by andrew-schofield**
* Set protocol handler for all platforms
* Add single instance checker, to allow forwarding of arguments to the original instance on windows
* Allow processing of monux:// auth url on windows

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -20,6 +20,11 @@ const mainWindow = new WindowManager()
 
 enableLiveReload()
 
+if(!app.isDefaultProtocolClient(app.getName())
```
Comment (robjtede) [inline src/app.ts]: For sanity, maybe add a `.toLowerCase()` here since `.getName()` returns capitalised "Monux".

File: src/app.ts [hunk 2]
```diff
@@ -49,12 +54,73 @@ const getAppInfo = (() => {
   }
 })()
 
+async function parseAuthUrl(forwardedUrl) {
+  const appInfo = await getAppInfo()
+  const query = parseUrl(forwardedUrl).query
+  const authResponse = parseQueryString(query)
+
+  if (authResponse.state !== appInfo.state) {
```
Comment (robjtede) [inline src/app.ts]: Add debug statement for the auth states in here.

File: src/app.ts [hunk 3]
```diff
@@ -20,6 +20,11 @@ const mainWindow = new WindowManager()
 
 enableLiveReload()
 
+if(!app.isDefaultProtocolClient(app.getName())
+{
+    app.setAsDefaultProtocolClient(app.getName())
```
Comment (robjtede) [inline src/app.ts]: Same as above here.

File: src/app.ts [hunk 4]
```diff
@@ -97,46 +163,14 @@ app.on('ready', async () => {
...
-    console.error(err)
-    throw new Error(err)
-  }
+  await parseAuthUrl(forwardedUrl)
 })
 
 app.on('window-all-closed', () => {
   debug('window-all-closed event')
 
   // conflicts with auth strategy for now
```
Comment (robjtede) [inline src/app.ts]: Remove extraneous comment.

File: src/app.ts [hunk 5]
```diff
@@ -49,12 +54,73 @@ const getAppInfo = (() => {
...
+  }
+}
+
+const isSecondInstance = app.makeSingleInstance(async (commandLine, workingDirectory) => {
+  // Someone tried to run a second instance, we should focus our window.
+  if (mainWindow) {
+    if (mainWindow.hasWindow) mainWindow.focus()
+    mainWindow.focus()
+    if( process.platform === 'win32' && commandLine.length > 1) {
+	  var urlParams = commandLine.filter(function(param){
```
Comment (robjtede) [inline src/app.ts]: `.filter` could be replaced with a `.find` and avoid the array accesses later.

Eg.

```
const authUrl = commandLine.filter((param) => {
  return param.startsWith('monux://')
})

if (authUrl) {
  await parseAuthUrl(urlParams[0])
} else {
  console.err('Auth url not found')
  throw new Error('Auth url not found')
}
```

Comment (robjtede) [inline src/app.ts]: I'd prefer arrow functions.

File: src/app.ts [hunk 6]
```diff
@@ -49,12 +54,73 @@ const getAppInfo = (() => {
   }
 })()
 
+async function parseAuthUrl(forwardedUrl) {
```
Comment (robjtede) [inline src/app.ts]: I'd prefer arrow functions.

`const parseAuthUrl = async (forwardedUrl) => {`

**Comment by robjtede** [comment]
I guess it should be rebased on develop to solve the merge conflicts as well.

Looks like the only conflict arises where:

```
await saveCode('refresh_token', refreshToken)
```

should now be:

```
if (refreshToken) await saveCode('refresh_token', refreshToken)
else debug('no refresh token sent')
```

File: src/app.ts (hunk 5 above)
Comment (andrew-schofield) [inline src/app.ts]: Not sure I understand this one, what should the arrow function be?

**Pull request merged**